pages/blog/_posts/banger.mdx

```diff
@@ -0,0 +1,279 @@
+      tempFilePath
+    ])
+
+    const croppedAudioS3Key = await getAudioFileKey(videoId)
+
+    // upload mp3 to file storage
+    const mp3URL = await uploadFile({
+      Key: croppedAudioS3Key,
+      Body: fs.createReadStream(tempFilePath)
+    })
+
+    // remove temp file
+    await unlink(tempFilePath)
+
+    await dbConnect()
+
+    await VideoModel.updateOne(
+      { _id: videoId },
+      { $set: { croppedAudioFileURL: mp3URL } }
+    )
+
+    return mp3URL
+  }
+)
+```
+
+### 3. Rendering the video
+
+The next step is to render the video using remote workers with beefy CPUs and GPUs.
+
+![A screenshot of the Banger app](/assets/blog/banger/banger-app.png)
+
+We have a sub-queue that communicates with our own infrastructure. We send a job to the queue while Inngest allows us to wait until the job is done and handles new progress events.
+
+```js
+const { videoFileURL, renderTime } = await step.run(
+  'Render video to s3',
+  async () => {
+    const outKey = await getVideoOutKey(videoId)
+
+    const userBundle = bundles.find((p) => p.key === updatedVideo.user.bundle)
+
+    if (!userBundle) {
+      throw new NonRetriableError('no bundle assigned to user')
+    }
+
+    await dbConnect()
+
+    const video = await VideoModel.findOne({
+      _id: videoId
+    }).populate('user')
+
+    if (!video) {
+      throw new NonRetriableError('no video found')
+    }
+
+    // attempt is provided by Inngest.
+    // if video fails to render from the first attempt, we will pick different worker
+    const renderer = await determineRenderer(video, attempt)
+
+    // CRF of the video based on user bundle
+    const constantRateFactor = determineRemotionConstantRateFactor(
+      video.user.bundle
+    )
+
+    const renderPriority = await determineQueuePriority(video.user.bundle)
+
+    logger.info(
+      `Rendering Remotion video with renderer ${renderer} and crf ${constantRateFactor}`
+    )
+
+    let processingStatusIsSet = false
+
+    const renderedVideo = await renderVideo({
+      videoId: videoId,
+      priority: renderPriority,
+      renderOptions: {
+        crf: constantRateFactor,
+        concurrency: determineRemotionConcurrency(video),
+        ...(video.hdr && {
+          colorSpace: 'bt2020-ncl'
+        })
+      },
+      inputPropsOverride: {
+        ...video.videoSettings,
+        videoFormat: video.videoFormat
+      },
+      renderer,
+      audioURL: croppedMp3Url,
+      startTime: 0,
+      endTime: video.videoDuration,
+      outKey,
+      onProgress: async (progress) => {
+        await VideoModel.updateOne(
+          {
+            _id: videoId
+          },
+          { $set: { renderProgress: progress, status: 'processing' } }
+        )
+      }
+    })
+
+    return renderedVideo
+  }
+)
+```
+
+### 4. Invalidate CDN cache
+
+There are some cases when video needs to be re-rendered. To make sure CDN cache is always fresh, we purge it each time the video renders.
```

Comment:
Extra context like "We host our videos on a CDN, but some times, the video needs to be re-rendered. To make sure..."